week03/inho/java-christmas-6/src/main/java/christmas/validator/MenuValidator.java

```diff
@@ -0,0 +1,120 @@
+package christmas.validator;
+
+
+import christmas.constant.ErrorMessage;
+import christmas.constant.menu.MenuKind;
+import christmas.model.Menu;
+
+import java.util.*;
+
+public class MenuValidator {
+
+    private LinkedHashMap<String, Integer> orders;
+    private List<Menu> menus;
+    private ErrorMessage errorMessage;
+    private static final int MIN_ORDER = 1;
+    private static final int MAX_ORDER = 20;
+    private static final int SPLIT_LIMIT = -1; // 실제 split 메서드 안에 두 번째 매개변수 이름
+
+    public boolean validateMenu(String otherString) {
+        orders = new LinkedHashMap<>();
+        menus = new ArrayList<>();
+
+        otherString = otherString.replaceAll(" ", "");
+
+        return validOtherString(otherString) && validContainUsingMenu(orders) &&
+                validLessMinOverMax(menus) && validOrderOnlyDrink(menus);
+    }
+
+    private boolean validOtherString(String otherString) {
+        try {
+            parsingAndSettingOrderString(otherString);
+        } catch (IllegalArgumentException e) {
+            errorMessage = ErrorMessage.ERROR_ORDER;
+
+            return false;
+        }
+        return true;
+    }
+
+    private void parsingAndSettingOrderString(String otherString) {
+        List<String> orderStrings = Arrays.stream(otherString.split(",", SPLIT_LIMIT)).toList();
+        orderStrings.forEach(this::settingOrderMap);
+    }
+
+    private void settingOrderMap(String orderString) {
+        String[] order = orderString.split("-", SPLIT_LIMIT);
+        int size = Integer.parseInt(order[1]);
+
+        if (size <= 0)
+            throw new IllegalArgumentException();
+
+        if (orders.containsKey(order[0]))
+            throw new IllegalArgumentException();
+
+        orders.put(order[0], size);
+    }
+
+    private boolean validContainUsingMenu(HashMap<String, Integer> orders) {
+        try {
+            orders.forEach((name, size) -> menus.add(new Menu(name, size)));
+        } catch (IllegalArgumentException e) {
+            errorMessage = ErrorMessage.ERROR_ORDER;
+
+            return false;
+        }
+        return true;
+    }
+
+    private boolean validLessMinOverMax(List<Menu> menus) {
+        int size = calculateTotalOrderSize(menus);
+
+        try {
+            if (size < MIN_ORDER || size > MAX_ORDER) {
+                throw new IllegalArgumentException();
+            }
+        } catch (IllegalArgumentException e) {
+            errorMessage = ErrorMessage.ERROR_OVER_MAX_SIZE;
+
+            return false;
+        }
+        return true;
+    }
```

Comment:
직접 생성한 예외를 던지고 잡아서 특정 로직을 추가한다면 if 문으로도 충분히 가능할 것 같아요.

```suggestion
    private boolean validOrderSizeWithinLimits(List<Menu> menus) {
        int size = calculateTotalOrderSize(menus);
 
        if (size < MIN_ORDER || size > MAX_ORDER) {
            errorMessage = ErrorMessage.ERROR_OVER_MAX_SIZE;
            return false;
        }

        return true;
    }
```

Reply:
아 저도 이런 느낌으로 구현했다가 에러를 발생시키고 잡는 로직으로 설계하라는 조건 때문에 try-catch를 사용했습니다!